units/src/amount/serde.rs

```diff
@@ -87,150 +52,78 @@ impl fmt::Display for DisplayFullError {
-    }
-    #[cfg(feature = "alloc")]
-    fn des_btc<'d, D: Deserializer<'d>>(d: D, _: private::Token) -> Result<Self, D::Error> {
-        use serde::de::Error;
-        SignedAmount::from_btc(f64::deserialize(d)?)
-            .map_err(DisplayFullError)
-            .map_err(D::Error::custom)
-    }
-    #[cfg(feature = "alloc")]
-    fn ser_str<S: Serializer>(self, s: S, _: private::Token) -> Result<S::Ok, S::Error> {
-        s.serialize_str(self.to_string_in(Denomination::Bitcoin).as_str())
-    }
-    #[cfg(feature = "alloc")]
-    fn des_str<'d, D: Deserializer<'d>>(d: D, _: private::Token) -> Result<Self, D::Error> {
-        use serde::de::Error;
-        let s: alloc::string::String = Deserialize::deserialize(d)?;
-        SignedAmount::from_str_in(&s, Denomination::Bitcoin)
-            .map_err(DisplayFullError)
-            .map_err(D::Error::custom)
-    }
-}
-
-impl SerdeAmountForOpt for SignedAmount {
-    fn type_prefix(_: private::Token) -> &'static str { "i" }
-    fn ser_sat_opt<S: Serializer>(self, s: S, _: private::Token) -> Result<S::Ok, S::Error> {
-        s.serialize_some(&self.to_sat())
-    }
-    #[cfg(feature = "alloc")]
-    fn ser_btc_opt<S: Serializer>(self, s: S, _: private::Token) -> Result<S::Ok, S::Error> {
-        s.serialize_some(&self.to_btc())
-    }
-    #[cfg(feature = "alloc")]
-    fn ser_str_opt<S: Serializer>(self, s: S, _: private::Token) -> Result<S::Ok, S::Error> {
-        s.serialize_some(&self.to_string_in(Denomination::Bitcoin))
-    }
-}
-
 pub mod as_sat {
-    //! Serialize and deserialize [`Amount`](crate::Amount) as real numbers denominated in satoshi.
+    //! Serialize and deserialize [`Amount`] and [`SignedAmount`] as real numbers denominated in satoshi.
+    //!
     //! Use with `#[serde(with = "amount::serde::as_sat")]`.
+    //!
+    //! [`Amount`]: crate::Amount
+    //! [`SignedAmount`]: crate::SignedAmount
 
-    use serde::{Deserializer, Serializer};
+    use serde::{Deserialize, Deserializer, Serialize, Serializer};
 
-    use super::private;
-    use crate::amount::serde::SerdeAmount;
+    use crate::SignedAmount;
 
-    pub fn serialize<A: SerdeAmount, S: Serializer>(a: &A, s: S) -> Result<S::Ok, S::Error> {
-        a.ser_sat(s, private::Token)
+    pub fn serialize<A, S: Serializer>(a: &A, s: S) -> Result<S::Ok, S::Error>
+    where
+        A: Into<SignedAmount> + Copy,
+    {
+        let amount: SignedAmount = (*a).into();
+        i64::serialize(&amount.to_sat(), s)
     }
 
-    pub fn deserialize<'d, A: SerdeAmount, D: Deserializer<'d>>(d: D) -> Result<A, D::Error> {
-        A::des_sat(d, private::Token)
+    pub fn deserialize<'d, A, D: Deserializer<'d>>(d: D) -> Result<A, D::Error>
+    where
+        A: TryFrom<SignedAmount>, <A as TryFrom<SignedAmount>>::Error: core::fmt::Display
+    {
+        let sat = i64::deserialize(d)?;
+        let amount = SignedAmount::from_sat(sat)
+            .map_err(serde::de::Error::custom)?;
```

Comment:
This should be `serde::de::Error::invalid_value`

Reply:
In practice this results in much worse error messages because you can't preserve the original error.